### PR TITLE
Saving CartBounty abandonment data for compatibility

### PR DESCRIPTION
**Pull request by angbpy**

### CODE REVIEW

File: includes/api/save-abandonment-data.php [hunk 1]
```diff
@@ -122,6 +122,16 @@ function saveCartAbandonmentData(WP_REST_Request $request)
     return new WP_REST_Response($response, $statusCode);
 }
 
+
+//Check CartBounty plugin is activated or not 
+if (is_plugin_active('woo-save-abandoned-carts/cartbounty-abandoned-carts.php') && empty($customerEmail) == false) {
+
```
Comment (ChetanGN) [inline includes/api/save-abandonment-data.php:128]: @angbpy Remove extra white spaces. Maintain the same code format.

Comment (angbpy) [reply includes/api/save-abandonment-data.php:127]: @ChetanGN Done. Please check.

File: includes/api/save-abandonment-data.php [hunk 2]
```diff
@@ -291,6 +298,546 @@ function saveWooCartAbandonmentRecoveryData($razorpayData)
...
+    
+    $result['response']    = $response;
+    $result['status_code'] = $statusCode;
+    return $result;
+}
+
+function getSessionID($order_id){
+  $session_id = WC()->session->get_customer_id();
+  $order      = wc_get_order( $order_id );
+  $user       = $order->get_user();
```
[the remark this answers is not in the document]
Comment (angbpy) [reply includes/api/save-abandonment-data.php:413]: Removed line #413

File: includes/api/save-abandonment-data.php [hunk 3]
```diff
@@ -291,6 +298,546 @@ function saveWooCartAbandonmentRecoveryData($razorpayData)
...
+  }
+  return $session_id;
+}
+
+function handle_order( $order_id ){
+    if( !isset($order_id) ){ //Exit if Order ID is not present
+        return;
+    }
+
+     $public = new CartBounty_Public(CARTBOUNTY_PLUGIN_NAME_SLUG, CARTBOUNTY_VERSION_NUMBER);
```
[the remark this answers is not in the document]
Comment (angbpy) [reply includes/api/save-abandonment-data.php:435]: Defined by CartBounty only, we are able to use it.

File: includes/api/save-abandonment-data.php [hunk 4]
```diff
@@ -291,6 +298,546 @@ function saveWooCartAbandonmentRecoveryData($razorpayData)
     return $logObj;
 }
 
+// ___________________ CARTBOUNTY PLUGIN FUNCTIONS _______________________________________________________________________________
```
[the remark this answers is not in the document]
Comment (angbpy) [reply includes/api/save-abandonment-data.php]: Removed Cartbounty label lines

File: includes/api/save-abandonment-data.php [hunk 5]
```diff
@@ -291,6 +298,433 @@ function saveWooCartAbandonmentRecoveryData($razorpayData)
...
+//Save cart abandonment data for CartBounty plugin
+function saveCartBountyData($razorpayData){
+    global $wpdb;
+    $products        = WC()->cart->get_cart();
+    $ghost           = false;
+    $name            = $razorpayData['customer_details']['billing_address']['name'] ?? '';
+    $surname         = " ";
+    $email           = $razorpayData['customer_details']['email'];
+    $phone           = $razorpayData['customer_details']['contact'];
+    $cart_table      = $wpdb->prefix ."cartbounty";
```
Comment (ChetanGN) [inline includes/api/save-abandonment-data.php:310]: @angbpy maintains variable names in camelcase.

File: includes/api/save-abandonment-data.php [hunk 6]
```diff
@@ -291,6 +298,433 @@ function saveWooCartAbandonmentRecoveryData($razorpayData)
...
+      'cartbounty_shipping_state' 		=> $razorpayData['customer_details']['shipping_address']['state'] ?? '',
+      'cartbounty_shipping_postcode' 	=> $razorpayData['customer_details']['shipping_address']['zipcode'] ?? '',
+      'cartbounty_order_comments' 	    => '',
+      'cartbounty_create_account' 	    => '',
+      'cartbounty_ship_elsewhere' 		=> ''
+  );
+
+  $user_data = array(
+      'name'			=> $name,
+      'surname'		    => $surname,
```
Comment (ChetanGN) [inline includes/api/save-abandonment-data.php:340]: @angbpy formatting the code.

Comment (angbpy) [reply includes/api/save-abandonment-data.php:340]: @ChetanGN This is actually correctly formatted , it looks this way on github

File: includes/api/save-abandonment-data.php [hunk 7]
```diff
@@ -291,6 +298,433 @@ function saveWooCartAbandonmentRecoveryData($razorpayData)
...
+          $session_id = WC()->session->get_customer_id(); //Retrieving customer ID from WooCommerce sessions variable
+         }
+       if( WC()->session->get( 'cartbounty_from_link' ) && WC()->session->get( 'cartbounty_session_id' ) ){
+          $session_id = WC()->session->get( 'cartbounty_session_id' );
+         }
+  }
+  return $session_id;
+}
+
+function handleOrder( $order_id ){
```
Comment (ChetanGN) [inline includes/api/save-abandonment-data.php:427]: @angbpy Function name should be `handleCBRecoveredOrder`

**Comment by ChetanGN** [comment]
@angbpy I think we need to move this plugin support functionality to a different file.

File: includes/api/save-abandonment-data.php (hunk 5 above)
Comment (angbpy) [reply includes/api/save-abandonment-data.php]: Done and updated

File: includes/api/save-abandonment-data.php (hunk 7 above)
Comment (angbpy) [reply includes/api/save-abandonment-data.php]: Done and updated

**Comment by angbpy** [comment]
> @angbpy I think we need to move this plugin support functionality to a different file.

Moved to different files and removed repeated functions.

File: woo-razorpay.php [hunk 8]
```diff
@@ -2019,3 +2019,10 @@ function disable_coupon_field_on_cart($enabled)
     add_action('woocommerce_cart_updated', 'enqueueScriptsFor1cc', 10);
     add_filter('woocommerce_order_needs_shipping_address', '__return_true');
 }
+
+//Changes Recovery link URL to Magic cart URL to avoid redirection to checkout page
+function cartbounty_alter_automation_button( $button ){
+    return str_replace("cartbounty=","cartbounty=magic_",$button);
+}
+
+add_filter( 'cartbounty_automation_button_html', 'cartbounty_alter_automation_button' );
```
Comment (ChetanGN) [inline woo-razorpay.php:2028]: What if this plugin is not installed? Will this hook throws an exception?

Comment (angbpy) [reply woo-razorpay.php:2028]: Should I add a check whether the plugin is active and put this within that?

File: includes/support/cartbounty.php [hunk 9]
```diff
@@ -0,0 +1,310 @@
...
+            sanitize_text_field($userData['name']),
+            sanitize_text_field($userData['surname']),
+            sanitize_email($userData['email']),
+            filter_var($userData['phone'], FILTER_SANITIZE_NUMBER_INT),
+            sanitize_text_field(serialize($userData['location'])),
+            $sessionID
+        )
+    );
+
+    deleteDuplicateCarts($sessionID, $updatedRows, $cartTable);
```
Comment (ChetanGN) [inline includes/support/cartbounty.php:98]: Still not using plugin function?

Comment (angbpy) [reply includes/support/cartbounty.php:98]: We can't use it as it is throwing critical errors.
This is a private function within it and the rest are public ones and we're able to reuse them.

**Pull request closed without merging**